### PR TITLE
Refactor FXIOS-7595 [v120] When there are no more highlights to display, do not show the "Show more" button

### PR DESCRIPTION
**Pull request by PARAIPAN9**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-7595)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/16899)

## Video

https://github.com/mozilla-mobile/firefox-ios/assets/51127880/9955df65-76ce-47a1-ad00-a1a06840a10b

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [x] Wrote unit tests and/or ensured the tests suite is passing
- [x] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [x] If needed I updated documentation / comments for complex code and public methods

### CODE REVIEW

File: Shared/Extensions/String+Extension.swift [hunk 1]
```diff
@@ -131,6 +131,22 @@ public extension String {
...
     }
 
+    /// Calculate and return the width of a string when rendered with a given font.
+    ///
+    /// - Parameters:
+    ///   - string: The input string for which you want to calculate the width.
+    ///   - font: The font that will be used to render the string.
+    ///
+    /// - Returns: The width of the string when rendered with the specified font.
+    func widthOfString(usingFont font: UIFont) -> CGFloat {
```
Comment (razvanlitianu) [inline Shared/Extensions/String+Extension.swift:141]: nit: `func width(with font: Font) -> CGFloat` is clear enough for me.

File: Client/Frontend/Fakespot/Views/FakespotHighlightsCardView.swift [hunk 2]
```diff
@@ -86,6 +95,12 @@ class FakespotHighlightsCardView: UIView, ThemeApplicable {
     private var viewModel: FakespotHighlightsCardViewModel?
     private var isShowingPreview = true
 
+    private var safeAreaEdgeInsets: UIEdgeInsets {
+        guard let firstScene = UIApplication.shared.connectedScenes.first as? UIWindowScene else { return UIEdgeInsets() }
```
Comment (razvanlitianu) [inline Client/Frontend/Fakespot/Views/FakespotHighlightsCardView.swift:99]: Can we move this as a  `UIApplication.shared` extension? Doesn't seem to be related to this view.

Comment (PARAIPAN9) [reply Client/Frontend/Fakespot/Views/FakespotHighlightsCardView.swift:99]: Yes, sure.

File: Client/Frontend/Fakespot/Views/FakespotHighlightsCardView.swift [hunk 3]
```diff
@@ -141,6 +150,36 @@ class FakespotHighlightsCardView: UIView, ThemeApplicable {
         dividerView.backgroundColor = theme.colors.borderPrimary
     }
 
+    override func layoutSubviews() {
+        super.layoutSubviews()
+        guard let viewModel else { return }
```
Comment (thatswinnie) [inline Client/Frontend/Fakespot/Views/FakespotHighlightsCardView.swift:155]: Please combine these 3 guards into one.

File: Client/Frontend/Fakespot/Views/FakespotHighlightsCardView.swift [hunk 4]
```diff
@@ -141,6 +150,36 @@ class FakespotHighlightsCardView: UIView, ThemeApplicable {
...
+
+        let highlightLabelWidth = longestReview.width(
+            with: DefaultDynamicFontHelper.preferredFont(
+                withTextStyle: .subheadline,
+                size: UX.titleFontSize,
+                weight: .semibold
+            )
+        )
+
+        let highlightsGroupViewWidth = (bounds.width - UIWindow.safeAreaInsets.left * 2) -
```
Comment (thatswinnie) [inline Client/Frontend/Fakespot/Views/FakespotHighlightsCardView.swift:167]: Please add a comment on this calculation. Also we shouldn't force unwrap anything here - you could add `highlightGroups.first` to the guards above. The magic number `24` should be replaced with a UX constant.

File: Client/Frontend/Fakespot/Views/FakespotHighlightsCardView.swift [hunk 5]
```diff
@@ -141,6 +150,36 @@ class FakespotHighlightsCardView: UIView, ThemeApplicable {
...
+        let areMoreThanTwoLines = highlightLabelWidth > highlightsGroupViewWidth
+
+        if areMoreThanTwoLines {
+            contentStackView.addArrangedSubview(dividerView)
+            contentStackView.addArrangedSubview(moreButton)
+        } else {
+            contentStackView.removeArrangedView(dividerView)
+            contentStackView.removeArrangedView(moreButton)
+        }
+        updateHighlights(areMoreThanTwoLines)
```
Comment (thatswinnie) [inline Client/Frontend/Fakespot/Views/FakespotHighlightsCardView.swift:180]: I think it would be better to extract all of the code added in `layoutSubviews` into a method.

File: Shared/Extensions/String+Extension.swift [hunk 6]
```diff
@@ -131,6 +131,20 @@ public extension String {
         return nsString.substring(with: match.range(at: 1))
     }
 
+    /// Calculates and returns the width of a string when rendered with a specified font.
+    ///
+    /// - Parameter font: The font that will be used to render the string.
+    ///
+    /// - Returns: The width of the string when rendered with the specified font.
+    func width(with font: UIFont) -> CGFloat {
```
Comment (lmarceau) [inline Shared/Extensions/String+Extension.swift:139]: IMO I would keep this in the FakeSpot VC. I don't think we want this to be a standard practice to measure the width of a string. It should be an exception

Comment (razvanlitianu) [reply Shared/Extensions/String+Extension.swift:139]: I get where Sorin's coming from. It's kind of a Swift thing to do, and sticking it in Fakespot VC could clutter things up with extras that aren't really about what the VC is supposed to do. Extensions are pretty handy in Swift for this sort of thing, so maybe that's the way to go.

Comment (lmarceau) [reply Shared/Extensions/String+Extension.swift:139]: It's a basic Swift thing to do yes and most people do it by default, but we challenged that in previous eng sync meetings. We also had this conversation multiple times together already, but as said in the past we're trying to avoid extension usage unless specifically needed. Measuring the width of a string isn't a behavior we should have by default as it encourage going away from supporting automatic sizing, which is needed for dynamic type. It can be done where needed, but I wouldn't vote to encourage such behavior in our app. Happy to discuss further if needed and bring this up to the team.

File: Client/Extensions/UIWindow+Extension.swift [hunk 7]
```diff
@@ -40,5 +40,10 @@ extension UIWindow {
     static var statusBarHeight: CGFloat {
         keyWindow?.windowScene?.statusBarManager?.statusBarFrame.height ?? 0
     }
+
+    static var safeAreaInsets: UIEdgeInsets {
```
Comment (thatswinnie) [inline Client/Extensions/UIWindow+Extension.swift:44]: This should be moved to `FakespotHighlightsCardView`.

**Review by thatswinnie — APPROVED** [review]

**Comment by thatswinnie** [comment]
@Mergifyio backport release/v120
